[Bulk User]: Handle bad-data errors w/ `handleApiError` (+ better side panel refresh handling)

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

- In the `apiResource.js` file, the `logError` function expected an error object to have a particular shape - but the error that this issue fixes was getting swallowed up by the errors thrown when the error object in `logError` wasn't the shape expected. I added some existence checks to ensure errors don't get annihilated.
- I noticed that refreshing w/ a side panel open would load the page w/ the side panel visible. I added checks in `onMounted` hooks to redirect to the root page (using the `goBack` module function) whenever there are no users selected.

This will give the user the ability to refresh the page, which will remove stale data.

## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

Fixes #13550 

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

- Refresh the page on each of the side panels
- Perform the replication steps noted in #13550 and see that when you get a bad-data error there, you are shown the API error page and that when you click "Refresh" you are taken to a fresh users table.

nucleogenesis committed:
more safely handle err.config properties in logError

nucleogenesis committed:
bulk user panels go to root table when no users selected (ie, on refresh)

pcenov commented:
Hi @nucleogenesis - I confirm that now if I try to enrol a user in an already deleted class then I am seeing the "Sorry! Something went wrong" screen and the user can refresh the page. 

Will you be handling the following similar edge cases in a similar manner within the scope of this PR or I should report them separately:

1. Trying to remove a user from a deleted class results in a success message (which is incorrect as the action was never completed) and consequently clicking the UNDO button results in an error message in the snackbar.:

https://github.com/user-attachments/assets/1f08a054-7d3a-486f-9a9b-d3a4ad301c01

I would expect to see the generic error page in this case too, allowing me to refresh.

2. Similarly when assigning coaches to a deleted class the user is stuck at the side panel:

https://github.com/user-attachments/assets/65c921e6-aa86-4ada-92ca-9618e5d155e7

3. When creating a new user and attempting to assign or enrol the user in the deleted class there's no indication to the user that this is no longer possible:

https://github.com/user-attachments/assets/c96d6fed-524f-4791-bd66-dfbe003d44d6

4. Actions performed on an already deleted user:

https://github.com/user-attachments/assets/1b69fb43-57f5-43ac-b621-d60166b563ea

5. Attempts to interact with a permanently deleted user:

https://github.com/user-attachments/assets/cd929827-569e-440d-93e0-dbd826d438eb

AlexVelezLl commented on kolibri/plugins/facility/assets/src/views/users/sidePanels/EnrollLearnersSidePanel.vue:
```diff
@@ -124,6 +124,23 @@
+        if (props.selectedUsers.size === 0) {
+          goBack();
+        }
+      });
```
Just reflecting if we should wait until the component is mounted to do this check? I think we can also prevent the load data API calls like this https://github.com/learningequality/kolibri/pull/13769/files#diff-1004c38374e2196707382df426d82066fb2c3e124faa549273ceff802b59695bR203 if the are no selected users.

nucleogenesis replied:
I think your link might be pointing to the wrong place it's just bringing me back to this comment. 

I did try it w/out the onMounted hook but it didn't work but if there's a better way I'll apply it here.

nucleogenesis replied:
I've ended up moving this into a `beforeRouteEnter` guard @AlexVelezLl so that should avoid unneeded API calls.

marcellamaki commented on kolibri/plugins/facility/assets/src/views/users/sidePanels/EnrollLearnersSidePanel.vue:
```diff
@@ -209,7 +217,8 @@
             createdMemberships.value = newMemberships;
           } catch (error) {
-            showErrorWarning.value = true;
+            // Why doesn't this ever get run here?
```
I don't know what this comment means?

nucleogenesis replied:
Thanks for catching this lol - it is an artifact of my confusion caused by the errors being swallowed up in the apiResource module :sweat_smile:

nucleogenesis committed:
handle refresh in nav guard, not onMounted, avoid whole setup() function this way

marcellamaki reviewed (commented):
the changes you just pushed do seem to simplify things, although having read the code, I'm not sure that I'm the best positioned to be a final approver, since it seems like both @AlexVelezLl and @rtibbles had some concrete recommendations/feedback

nucleogenesis committed:
errant comment

rtibbles reviewed (commented):
I had far fewer opinions than I thought I would. This is fine to go as long as manual testing of these specific points check out (others will be filed in follow ups).

marcellamaki reviewed (approved):
Refresh the page on each of the side panels ✅ 
Perform the replication steps noted in https://github.com/learningequality/kolibri/issues/13550 and see that when you get a bad-data error there, you are shown the API error page and that when you click "Refresh" you are taken to a fresh users table. ✅ 

manual testing confirms both of these. I will leave more robust edge case finding to peter in our beta0 :) good to go

AlexVelezLl reviewed (commented):
Just flagging that the implementation of the `beforeRouteEnter` is partially incorrect, because these side panels can also be opened from the new users page. So redirecting them to the main user management page feel a bit weird? It is an edge case though, so will let you decide if this is a blocker or not!

https://github.com/user-attachments/assets/efa6c743-ef80-4857-8a6f-4cd0ff1361b3

AlexVelezLl commented:
In general I think that it does not makes much sense of having these side panels as route components. IIRC, the reason for having the side panels as route components was to respond to the question of "what happens if the user refresh the page?" and the reason of using routes for displaying the side panels was to keep the side panel open after the user refresh. But in this case this does not apply well because the user selection gets removed when they refresh. And now with this `beforeRouteLeave` we are basically just removing the reason why we were using routes for displaying the side panels. 


Having these side panels coming from different parent routes have made the implementation of the routes definition, the dependencies definition of the components (as all of them are hidden behind the `router-view` tag), and the `close side panel` handling a bit more complex, and I think we are not getting anything good from having them as route components.

An option for making it worth could be exploring saving the selected users state in the local storage or route query params, so when the user reloads they keep seeing the side panels with the previously selected users, but I think that this will also add some more complexity to the implementation.

AlexVelezLl reviewed (changes requested):
Adding a blocking review until we figure out how to proceed regarding the previous review.

nucleogenesis commented:
Thanks Alex - I'll definitely apply the changes to ensure the panels all work on the new/removed users pages.

I agree re: the routing - but I may put that into a follow-up issue.

nucleogenesis committed:
oops duped in rebase

nucleogenesis committed:
Removing users from deleted class - no need to show error message here about the class

The problem with showing an error here is that the removal can be
partially successful. The error doesn't show up when removing, but
rather when undoing the removal - these changes simply proceed along as
if everything went okay because ... you're removing users from the class
and, technically, they've already been removed from the class :)

AlexVelezLl commented on kolibri/plugins/facility/assets/src/views/users/sidePanels/EnrollLearnersSidePanel.vue:
```diff
@@ -223,7 +225,7 @@
             createdMemberships.value = newMemberships;
           } catch (error) {
-            showErrorWarning.value = true;
+            store.dispatch('handleApiError', { error });
```
Should this specific side panel have a different error handling on its main action? The other side panels handle this by setting `showErrorWarning` to true.

nucleogenesis replied:
I understood the issue Peter reported was that the error warning was unhelpful in this case because knowing something went wrong is half the battle - the other issue is that the most likely solution is to refresh the page & try again so dispatching the API error here ensures that happens.

To be honest, though, I think a more robust errors reporting would be ideal here to let the user know specifically what actually happened (ie, "The class you selected 'Class Name' no longer exists"). I'll think on this a bit more when I spin this up again in a bit. Thanks Alex

AlexVelezLl commented on kolibri/plugins/facility/assets/src/views/users/sidePanels/RemoveFromClassSidePanel.vue:
```diff
@@ -267,14 +268,20 @@
-        ]);
+        try {
+          await Promise.all([
+            enrollments.length
+              ? MembershipResource.saveCollection({ data: enrollments })
+              : Promise.resolve(),
+            assignments.length
+              ? RoleResource.saveCollection({ data: assignments })
+              : Promise.resolve(),
+          ]);
+        } catch (_) {
+          showErrorWarning.value = true;
+        } finally {
+          loading.value = false;
+        }
```
The `undoUserRemoval` function is executed after the `RemoveFromClassSidePanel` has been unmounted, since this callback is executed in the undo snackbar, after the modal has been closed. Therefore, setting the `showErrorWarning` or `loading` refs here does not have any effect. Perhaps it would be a better idea to show this error in another snackbar?

nucleogenesis replied:
Pushed a change using a Snackbar - great catch Alex thank you!

AlexVelezLl reviewed (commented):
Just some thoughts!

nucleogenesis committed:
show undoing error message in SnackBar

AlexVelezLl reviewed (approved):
Code changes look good to me! Thanks @nucleogenesis!
